make python rpc handler to be singleton class

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25742 make python rpc handler to be singleton class**

python rpc handler right now is namespace + global variable, changed it to be singleton class as it can gurantee deterministic order of variable destruction. for namespace + global variable, we hit a process exit crash issue because global variables have dependencies and they are not destructed as expected

Differential Revision: [D17097999](https://our.internmc.facebook.com/intern/diff/D17097999/)

zhaojuanmao committed:
make python rpc handler to be singleton class

python rpc handler right now is namespace + global variable, changed it to be singleton class as it can gurantee deterministic order of variable destruction. for namespace + global variable, we hit a process exit crash issue because global variables have dependencies and they are not destructed as expected

Differential Revision: [D17097999](https://our.internmc.facebook.com/intern/diff/D17097999/)

pietern commented:
@zhaojuanmao Is there a dependency on the multiprocessing PR here? If not then this can be unstacked and merged separately/earlier.

zhaojuanmao commented:
@pietern originally planned to stack on top of spawn diff is to make sure tests still pass if we switch to spawn.  but I think right now it does not matter, we can get this diff in first and make sure spawn diff passed all tests after this diff in:)

will unstack this diff

zhaojuanmao committed:
Update on "make python rpc handler to be singleton class"


python rpc handler right now is namespace + global variable, changed it to be singleton class as it can gurantee deterministic order of variable destruction. for namespace + global variable, we hit a process exit crash issue because global variables have dependencies and they are not destructed as expected

Differential Revision: [D17097999](https://our.internmc.facebook.com/intern/diff/D17097999/)

zhaojuanmao committed:
Update on "make python rpc handler to be singleton class"


python rpc handler right now is namespace + global variable, changed it to be singleton class as it can gurantee deterministic order of variable destruction. for namespace + global variable, we hit a process exit crash issue because global variables have dependencies and they are not destructed as expected

Differential Revision: [D17097999](https://our.internmc.facebook.com/intern/diff/D17097999/)

zhaojuanmao committed:
Update on "make python rpc handler to be singleton class"


python rpc handler right now is namespace + global variable, changed it to be singleton class as it can gurantee deterministic order of variable destruction. for namespace + global variable, we hit a process exit crash issue because global variables have dependencies and they are not destructed as expected

Differential Revision: [D17097999](https://our.internmc.facebook.com/intern/diff/D17097999/)

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -7,16 +7,33 @@ namespace torch {
+// The singleton object is constructed at first when RPC agent is
+// constructed, where the python function in
+// torch/distributed/internal_rpc_utils.py are imported only once.
+class PYBIND11_EXPORT PythonRpcHandler {
```
Question: what do we need `PYBIND11_EXPORT` here?

zhaojuanmao replied:
otherwise, it will have warning/error "PythonRpc hander declared with greater visibility than the type of its field", this is know issue when including pybind11 types such as py::object https://fburl.com/0b0gobwx, adding this can fix this warning/error in OSS

mrshenli replied:
@gchanan @ezyang 

It seems this is the first time we use `PYBIND11_EXPORT`. Do we have our own version of `PYBIND11_EXPORT`, or is it OK to use it here?

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -7,16 +7,33 @@ namespace torch {
+  // Execute python UDF, result is pickled to binary string
+  std::vector<char> generatePythonUDFResult(const Message& request);
+  // Returned python UDF result is pickled binary string, so run python
+  // function to unpickle the python UDF result and return pyObject to user
```
nit: `pyObject` -> `py::object`

mrshenli reviewed (approved):
LGTM! Please make sure CI tests pass before landing.

mrshenli commented:
@pytorchbot retest this please

zhaojuanmao committed:
Update on "make python rpc handler to be singleton class"


python rpc handler right now is namespace + global variable, changed it to be singleton class as it can gurantee deterministic order of variable destruction. for namespace + global variable, we hit a process exit crash issue because global variables have dependencies and they are not destructed as expected

Differential Revision: [D17097999](https://our.internmc.facebook.com/intern/diff/D17097999/)
